snapcast: add snapserver and snapclient packages

xabolcs opened the pull request:
Maintainer: me / @xabolcs 
Compile tested: not yet
Run tested: not yet

Description:
Fixes: #23924

davidandreoletti commented:
@xabolcs  I am interested in this and already compile snapcast (server) on my end with OpenWRT snapshot. What's missing from your code (and PR) to get merged ?

xabolcs commented:
Hi!

Thaks for testing!

A few things:
* compile tested on ... ?
* runtime tested on ... ?
* ipk sizes?
* apk compatibility? (e.g. version number)

Adding myself as amaintainer would also help.

davidandreoletti commented:
On my end:
- Compiled snapserver (aarch64) binary on x86-64 machine. 
- Runtime tested on NanoPi R4s (aarch64)
- ipk sizes (snapserver 0.28.0): 743k 
- apk compatibility? I have not built any APK for OpenWRT so far :-)

xabolcs commented:
Sorry for noob question, but how did you build my branch?

davidandreoletti commented:
> Sorry for noob question, but how did you build my branch?

I did not build your branch. I thought you were asking about my hardware setup for an eventual test.

xabolcs commented:
OK, thanks for clearing up the misunderstanding! 🙏

I'll let you know when I have packages for you - built from this PR.

davidandreoletti commented:
> OK, thanks for clearing up the misunderstanding! 🙏
> 
> I'll let you know when I have packages for you - built from this PR.

(ping @xabolcs )

xabolcs commented:
Hi @davidandreoletti !

Thanks for the reminder! Sorry, I was carried away! :slightly_smiling_face: 

Well in the meantime Snapcast 0.29.0 came out which has HTTPS support, but I'm so noob I don't know how to add those dependencies! :see_no_evil: 
I mean:
```
Package snapserver is missing dependencies for the following libraries:
libcrypto.so.3
libssl.so.3
```

Sure, I could stick with 0.28.0. :shrug:

davidandreoletti commented:
Stick with 0.28 for this PR. 

0.29 dependencies will be investigated in a new PR.

xabolcs commented:
> Stick with 0.28 for this PR.

I had to backport a few Boost v1.87.0 compatibility commit for that ... sure its far more easier. :ok_hand:

xabolcs commented:
After backporting 3 commits, 0.28.0 is compiling again! :tada: 

Here is the result for SNAPSHOT (and `apk`), please test!:
[snapcast-apk.tar.gz](https://github.com/user-attachments/files/18460338/snapcast-apk.tar.gz)

xabolcs commented:
Almost there:
> Dirty patches detected, please refresh and review the diff
Error: Process completed with exit code 1.

davidandreoletti commented:
@xabolcs My build system broke with the switch to OpenWRT's new "apk" package manager. I can compile this PR's snapcast package now. Will get back to you in a few days with my testing on (the same) hardware as you.

davidandreoletti commented:
@xabolcs Snapcast works on NanoPi R4s (4G) with OpenWRT master + this PR.

Thank you!

davidandreoletti commented:
@neheb This is a new package. Both individuals confirm the package compile and works on their machine. 
What's the next step to get this merged ?

neheb commented on sound/snapcast/Makefile:
```diff
@@ -0,0 +1,127 @@
+PKG_NAME:=snapcast
+BASE_VERSION:=0.28.0
+PKG_SOURCE_DATE:=2024-04-15
+PKG_SOURCE_VERSION:=5fa01b8a49d842ed368138ee071cdf1c34be7a64
```
why not just use the version here? something like

```
PKG_VERSION:=0.28.0
PKG_SOURCE_VERSION:=v$(PKG_VERSION)
```

as is done elsewhere.

xabolcs replied:
> why not just use the version here?

[Snapcast has](https://github.com/badaix/snapcast/branches) `master` branch for release versions and `develop` branch for nightly builds.

In the older days there was too much commit between the two and the development version was stable enough to use it.

That's why I wanted to include the exact commit and commit date from what the actual version was built.

Sure, **badaix** releases Snapcast often these days, so your suggestion could work too.
I would stick with this advanced versioning, but if you ask me to change it to the simple versioning, I would do it.

neheb replied:
I'd like that. If needed, SOURCE_DATE can be revisited.

neheb commented on sound/snapcast/Makefile:
```diff
@@ -0,0 +1,127 @@
+	SECTION := sound
+	CATEGORY := Sound
+	TITLE := Synchronous multiroom audio player
+	DEPENDS := +AUDIO_SUPPORT:alsa-lib +libstdcpp +libavahi-client +libatomic +libogg +libflac +libopus +boost +libsoxr
```
libstdcpp is included with boost. Minor detail.

xabolcs replied:
> libstdcpp is included with boost. Minor detail.

Thank you very much! I'm a package maintainer newbie.

neheb commented:
@davidandreoletti unfortunately the answer is constant pinging and hoping someone cares.

davidandreoletti commented:
> @davidandreoletti unfortunately the answer is constant pinging and hoping someone cares.

For this package, @xabolcs  and I are active. Your 2 feedbacks will be acted on.
Are you ok @neheb ?

neheb commented:
huh?

davidandreoletti commented:
@neheb Your message sounded a bit negative: "unfortunately the answer is constant pinging" and perhaps you growing tired of "hey I need you" kind of pings.

Looks like I am misinterpreted. I am sorry and let's move on.

xabolcs commented:
> For this package, **xabolcs** and I are active.

May I add you to `PKG_MAINTAINER:=` as a co-maintainer? I could deal with the code (:sweat_smile:), but being active here (e.g. to watch the issues) and the extra testing would be a great help to me!

xabolcs commented:
>  snapcast: add Snapcast from badaix/snapos as is 

Adding a commit like this (with some [`Co-authored-by:`](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors)) does have a value?

davidandreoletti commented:
> > For this package, **xabolcs** and I are active.
> 
> 
> 
> @davidandreoletti 
> 
> May I add you to `PKG_MAINTAINER:=` as a co-maintainer? I could deal with the code ( :sweat_smile: ), but being active here (e.g. to watch the issues) and the extra testing would be a great help to me!

Yes.

xabolcs commented:
@davidandreoletti , PR updated, maybe the `PKG_MIRROR_HASH` is correct now.
Compiled successfully, please test it on your NanoPi!

xabolcs commented:
@davidandreoletti 

If I try to make Avahi dependency optional (incorporating badaix/snapos#4), I got the followig:
```
Package snapserver is missing dependencies for the following libraries:
libavahi-client.so.3
libavahi-common.so.3
libexpat.so.1
```

Any hints? :slightly_smiling_face:

davidandreoletti commented:
> Package snapserver is missing dependencies for the following libraries:

Is this an error when installing the packages ?

xabolcs commented:
> Is this an error when installing the packages ?

It happened when compiling.

But leave that to an another issue!

davidandreoletti commented:
> make image PROFILE=generic 'PACKAGES=luci-base luci-ssl-nginx luci-mod-admin-full luci-theme-bootstrap kmod-usb-storage kmod-usb-ohci kmod-usb-uhci debootstrap debian-archive-keyring gpgv2 ip-full rsync iperf3 htop vim-full vim-runtime ethtool pciutils lsblk blkid diffutils openssh-server openssh-sftp-server wget curl ca-certificates ddns-scripts luci-app-ddns etherwake luci-app-wol luci-proto-wireguard wireguard-tools tmux collectd-mod-thermal collectd-mod-cpu collectd-mod-disk collectd-mod-memory collectd collectd-mod-rrdtool luci-app-statistics jq yq adguardhome luci-app-sqm openssh-client autossh nut-web-cgi nut-upslog nut-upsc nut-upsmon nut-server nut-common nut block-mount e2fsprogs kmod-fs-ext4 kmod-usb-storage kmod-usb-storage-uas kmod-usb2 kmod-usb3 usbutils docker docker-compose dockerd coreutils-dd coreutils-sha256sum findutils-find eject gzip mosquitto shairport-sync-openssl mosquitto snapserver owntone shadow-useradd mpd-full ffmpeg flock mpc mosquitto-client procps-ng-pgrep procps-ng-kill wget wget-ssl qemu-x86_64-softmmu qemu-aarch64-softmmu qemu-ga qemu-bridge-helper qemu-img kmod-tun kmod-vhost-net telnet-bsd xz xz-utils curl cloudflared luci-app-cloudflared tailscale tailscaled iptables-nft kmod-ipt-conntrack kmod-ipt-conntrack-extra kmod-ipt-conntrack-label kmod-nft-nat kmod-ipt-nat libblkid' FILES=/var/tmp/openwrt/build/image/anthonya-router-hw0-dev/generic/files EXTRA_IMAGE_NAME=anthonya-router-hw0-dev BIN_DIR=/mnt/share_openwrt/image/releases/image/anthonya-router-hw0-dev/snapshots

```
Generate local signing keys...
WARNING: can't open config file: /builder/shared-workdir/build/staging_dir/host/etc/ssl/openssl.cnf
WARNING: can't open config file: /builder/shared-workdir/build/staging_dir/host/etc/ssl/openssl.cnf
read EC key
writing EC key
Checking 'true'... ok.
Checking 'false'... ok.
Checking 'working-make'... ok.
Checking 'case-sensitive-fs'... ok.
Checking 'proper-umask'... ok.
Checking 'perl-data-dumper'... ok.
Checking 'perl-findbin'... ok.
Checking 'perl-file-copy'... ok.
Checking 'perl-file-compare'... ok.
Checking 'perl-thread-queue'... ok.
Checking 'perl-ipc-cmd'... ok.
Checking 'tar'... ok.
Checking 'find'... ok.
Checking 'bash'... updated.
Checking 'xargs'... ok.
Checking 'patch'... ok.
Checking 'diff'... ok.
Checking 'cp'... updated.
Checking 'seq'... ok.
Checking 'awk'... ok.
Checking 'grep'... updated.
Checking 'egrep'... updated.
Checking 'getopt'... ok.
Checking 'realpath'... ok.
Checking 'stat'... ok.
Checking 'gzip'... updated.
Checking 'unzip'... ok.
Checking 'bzip2'... ok.
Checking 'wget'... ok.
Checking 'install'... ok.
Checking 'perl'... ok.
Checking 'python'... updated.
Checking 'python3'... updated.
Checking 'python3-distutils'... ok.
Checking 'python3-stdlib'... ok.
Checking 'file'... ok.
Checking 'which'... ok.
Checking 'ldconfig-stub'... ok.
Building images for x86 - Generic x86/64
Packages: luci-base luci-ssl-nginx luci-mod-admin-full luci-theme-bootstrap kmod-usb-storage kmod-usb-ohci kmod-usb-uhci debootstrap debian-archive-keyring gpgv2 ip-full rsync iperf3 htop vim-full vim-runtime ethtool pciutils lsblk blkid diffutils openssh-server openssh-sftp-server wget curl ca-certificates ddns-scripts luci-app-ddns etherwake luci-app-wol luci-proto-wireguard wireguard-tools tmux collectd-mod-thermal collectd-mod-cpu collectd-mod-disk collectd-mod-memory collectd collectd-mod-rrdtool luci-app-statistics jq yq adguardhome luci-app-sqm openssh-client autossh nut-web-cgi nut-upslog nut-upsc nut-upsmon nut-server nut-common nut block-mount e2fsprogs kmod-fs-ext4 kmod-usb-storage kmod-usb-storage-uas kmod-usb2 kmod-usb3 usbutils docker docker-compose dockerd coreutils-dd coreutils-sha256sum findutils-find eject gzip mosquitto shairport-sync-openssl mosquitto snapserver owntone shadow-useradd mpd-full ffmpeg flock mpc mosquitto-client procps-ng-pgrep procps-ng-kill wget wget-ssl qemu-x86_64-softmmu qemu-aarch64-softmmu qemu-ga qemu-bridge-helper qemu-img kmod-tun kmod-vhost-net telnet-bsd xz xz-utils curl cloudflared luci-app-cloudflared tailscale tailscaled iptables-nft kmod-ipt-conntrack kmod-ipt-conntrack-extra kmod-ipt-conntrack-label kmod-nft-nat kmod-ipt-nat libblkid apk-mbedtls base-files ca-bundle dnsmasq dropbear e2fsprogs firewall4 fstools grub2-bios-setup kernel kmod-amazon-ena kmod-amd-xgbe kmod-bnx2 kmod-button-hotplug kmod-drm-i915 kmod-dwmac-intel kmod-e1000 kmod-e1000e kmod-forcedeth kmod-fs-vfat kmod-igb kmod-igc kmod-ixgbe kmod-nft-offload kmod-r8169 kmod-tg3 libc libgcc libustream-mbedtls logd mkf2fs mtd netifd nftables odhcp6c odhcpd-ipv6only partx-utils ppp ppp-mod-pppoe procd-ujail uci uclient-fetch urandom-seed urngd

WARNING: /var/tmp/openwrt/build/image/anthonya-router-hw0-dev/generic/image/openwrt-imagebuilder/repositories:9: unrecogized keyword: packages/packages.adb
ERROR: failed to read repositories: /var/tmp/openwrt/build/image/anthonya-router-hw0-dev/generic/image/openwrt-imagebuilder/repositories: unsupported repositories file keyword
OK: 0 MiB in 0 packages
Package list missing or not up-to-date, generating it.

Building package index...

Installing packages...
WARNING: /var/tmp/openwrt/build/image/anthonya-router-hw0-dev/generic/image/openwrt-imagebuilder/repositories:9: unrecogized keyword: packages/packages.adb
ERROR: failed to read repositories: /var/tmp/openwrt/build/image/anthonya-router-hw0-dev/generic/image/openwrt-imagebuilder/repositories: unsupported repositories file keyword
ERROR: unable to select packages:
  snapserver (no such package):
    required by: world[snapserver]
make[2]: *** [Makefile:225: package_install] Error 1
make[1]: *** [Makefile:156: _call_image] Error 2
make: *** [Makefile:331: image] Error 2
```

This is what I get now building with the snapcast @ https://github.com/openwrt/packages/pull/23956/commits/eb3295a208017df640797369b1cbb8a3b7128e39 + https://github.com/openwrt/openwrt/commit/edfe58981193a424dfb2e636d7b97a88fff06743

davidandreoletti commented:
@xabolcs 
![image](https://github.com/user-attachments/assets/1473ea3d-2bc0-45be-b7e2-f8ec9c26bf55)

Have you tried rebasing your PR with openwrt's master branch to get the latest changes ?

xabolcs commented:
Yeah, I saw that this PR got outdated.
Will update in the next days.

Thanks for the ping!

azrdev commented:
I tried unsuccessfully to get a working snapclient.ipkg for my https://openwrt.org/toh/d-link/dir-860l_b1  using the openwrt "SDK": https://github.com/badaix/snapos/compare/master...azrdev:snapos:master

I'd be glad to test your build (even moreso an eventually merged package)

xabolcs commented:
> I tried unsuccessfully to get a working snapclient.ipkg for my DIR-860L B1.

I hope you run SNAPSHOT version of OpenWrt (with `apk` as package manager) on your awesome D-Link device!

After the rebase I could provide you an installable binary of Snapcast v0.28.0.

azrdev commented:
> I hope you run SNAPSHOT version of OpenWrt (with apk as package manager) on your awesome D-Link device!

Well yes, that's probably (part of) the issue: I run a freifunk firmware, which is based on https://github.com/freifunk-gluon/gluon/ and so I don't really know nor control what openwrt upstream revision this is based on :/

davidandreoletti commented:
> Yeah, I saw that this PR got outdated. Will update in the next days.
> 
> Thanks for the ping!

@xabolcs (ping) :-)

xabolcs commented:
> **xabolcs** (ping) :-)

pong!

Updated and removed draft flag.


@azrdev I could provide you a binary for your DIR-860L B1 with Snapcast server + client integrated.
Are you interested?

hunterzero99 commented:
I was able to build and run on 24.10.0.  Hardware is an RPi 4B

[snapclient_0.28.0-r1_aarch64_cortex-a72.ipk.zip](https://github.com/user-attachments/files/19167675/snapclient_0.28.0-r1_aarch64_cortex-a72.ipk.zip)
[snapserver_0.28.0-r1_aarch64_cortex-a72.ipk.zip](https://github.com/user-attachments/files/19167676/snapserver_0.28.0-r1_aarch64_cortex-a72.ipk.zip)

xabolcs commented:
> What's missing to merge this PR ?

This PR is ready now, again!
The only missing thing is a package manager's precious private time!

davidandreoletti commented:
> > What's missing to merge this PR ?
> 
> This PR is ready now, again! The only missing thing is a package manager's precious private time!

It was settled earlier: https://github.com/openwrt/packages/pull/23956#issuecomment-2650363472

xabolcs commented:
> It was settled earlier:

Ahh I meant the `openwrt-packages` maintainers, those who have commit access to this repository. :sweat_smile:

davidandreoletti commented:
> > It was settled earlier:
> 
> Ahh I meant the `openwrt-packages` maintainers, those who have commit access to this repository. 😅

Got it. 

@neheb: Would you mind merging this PR please ?

neheb commented:
maybe @1715173329 could merge.

1715173329 commented on sound/snapcast/Makefile:
```diff
@@ -0,0 +1,126 @@
+
+PKG_NAME:=snapcast
+PKG_VERSION:=0.28.0
+PKG_SOURCE_VERSION:=v$(PKG_VERSION)
+PKG_RELEASE:=1
+
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_URL:=https://github.com/badaix/snapcast.git
+PKG_MIRROR_HASH=288325fc8b134502adb522711c18210794734675684cf5d8eba8a0a82778f6c9
```
```suggestion
PKG_RELEASE:=1

PKG_SOURCE_PROTO:=git
PKG_SOURCE_URL:=https://github.com/badaix/snapcast.git
PKG_SOURCE_VERSION:=v$(PKG_VERSION)
PKG_MIRROR_HASH=288325fc8b134502adb522711c18210794734675684cf5d8eba8a0a82778f6c9
```

xabolcs replied:
Accepted suggested change!

1715173329 commented on sound/snapcast/Makefile:
```diff
@@ -0,0 +1,126 @@
+PKG_LICENSE_FILES:=LICENSE
+PKG_CPE_ID:=cpe:/a:badaix:snapcast
+
+PKG_BUILD_PARALLEL:=1
+
```
```suggestion
```

parallel build is enabled by default for cmake.

xabolcs replied:
Accepted suggested change!

1715173329 commented on sound/snapcast/Makefile:
```diff
@@ -0,0 +1,126 @@
+
+define Package/snapcast/description/Default
+	Synchronous audio player
+endef
+
+define Package/snapcast
+	$(call Package/snapcast/Default)
+	TITLE += packages
+endef
+
+define Package/snapcast/description
+	$(call Package/snapcast/description/Default)
+	Snapcast is a multi-room client-server audio player,
+	where all clients are time synchronized with the server
+	to play perfectly synced audio
+endef
+
+define Package/snapserver
+	$(call Package/snapcast/Default)
+	TITLE += Snapserver
+	DEPENDS += +libvorbis
+	HIDDEN := 1
+endef
+
+define Package/snapclient
+	$(call Package/snapcast/Default)
+	TILE += Snapclient
+	DEPENDS += +libvorbisidec
+	HIDDEN := 1
+endef
+
+define Package/snapserver/description
+	$(call Package/snapcast/description/Default)
+	Snapcast server
+endef
+
+define Package/snapclient/description
+	$(call Package/snapcast/description/Default)
+	Snapcast client
+endef
```
```suggestion
define Package/snapcast/Default
  SECTION:=sound
  CATEGORY:=Sound
  TITLE:=Synchronous multiroom audio player
  DEPENDS:=+AUDIO_SUPPORT:alsa-lib +libavahi-client +libatomic +libogg +libflac +libopus +boost +libsoxr
  URL:=https://github.com/badaix/snapcast
endef

define Package/snapcast/description/Default
  Synchronous audio player
endef

define Package/snapcast
  $(call Package/snapcast/Default)
  TITLE += packages
endef

define Package/snapcast/description
  $(call Package/snapcast/description/Default)
  Snapcast is a multi-room client-server audio player,
  where all clients are time synchronized with the server
  to play perfectly synced audio
endef

define Package/snapserver
  $(call Package/snapcast/Default)
  TITLE += Snapserver
  DEPENDS += +libvorbis
  HIDDEN:=1
endef

define Package/snapclient
  $(call Package/snapcast/Default)
  TITLE += Snapclient
  DEPENDS += +libvorbisidec
  HIDDEN:=1
endef

define Package/snapserver/description
  $(call Package/snapcast/description/Default)
  Snapcast server
endef

define Package/snapclient/description
  $(call Package/snapcast/description/Default)
  Snapcast client
endef
```

use two spaces for indentation and remove extra spaces between `:=`

xabolcs replied:
Accepted suggested change!

Did you left out the `install` sections from those "two spaces for indentation" suggestion out intentionally?

1715173329 replied:
> Did you left out the install sections from those "two spaces for indentation" suggestion out intentionally?

Yes, these are commands, not variable.

1715173329 commented on sound/snapcast/Makefile:
```diff
@@ -0,0 +1,126 @@
+/etc/default/snapclient
+endef
+
+define Package/snapcast/Default/install
```
this seems to do nothing, drop

xabolcs replied:
Accepted suggested change!

1715173329 commented on sound/snapcast/files/snapclient.init:
```diff
@@ -0,0 +1,31 @@
+NAME=snapclient
+PROG=/usr/bin/$NAME
+PID_FILE=/var/run/$NAME.pid
+CONFIG_FILE=/etc/default/$NAME
```
Please convert it to use UCI config. You may take [v2ray-core](https://github.com/openwrt/packages/tree/master/net/v2ray-core/files) as an example.

davidandreoletti replied:
@xabolcs To get started with / finish the PR, UCI config can offer a single option to a native snapserver or snapclient config file like in the example:

```
config snapserver 'config'
	option conf '/etc/default/snapserver'
```

As time passes, common uses cases could get better integration via new UCI options like stream sources, etc

xabolcs replied:
@davidandreoletti , I think that won't do it: I mean the review is about to avoid some new config parsing logic as there is already one.
I'm already rebasing my branch and implementing this, just have no time to finish. :see_no_evil: 
So I have some work-in-progress commit, but the rebase itself isn't finished yet, so I can't publish it.

Please be patient, I will try to find time to finish this.

davidandreoletti replied:
@xabolcs Hi, where you able to find some time eventually ?

xabolcs replied:
Addressed comment about UCI config! Thanks again for the pointer! :pray:

1715173329 commented on sound/snapcast/files/snapclient.init:
```diff
@@ -0,0 +1,31 @@
+PID_FILE=/var/run/$NAME.pid
+CONFIG_FILE=/etc/default/$NAME
+
+
```
```suggestion
```

remove extra blank line

xabolcs replied:
Accepted suggested change!

xabolcs commented:
>  **1715173329** requested changes

Thank you for reviewing and for the lead!

xabolcs commented:
@1715173329 

The `.conf` and `.init` files for `snapclient` and `snapserver` differs only a little bit:
```bash
$ diff -U8 files/snapclient.conf files/snapserver.conf 
--- files/snapclient.conf	2025-05-03 10:52:55.416510137 +0200
+++ files/snapserver.conf	2025-05-03 10:52:55.416510137 +0200
@@ -1,4 +1,4 @@
-config snapclient 'config'
+config snapserver 'config'
 	option autostart '0'
 	option log_sink 'system'
 	option opts ''

$ diff -U8 files/snapclient.init files/snapserver.init 
--- files/snapclient.init	2025-05-03 10:52:55.416510137 +0200
+++ files/snapserver.init	2025-05-03 10:52:55.416510137 +0200
@@ -1,17 +1,17 @@
 #!/bin/sh /etc/rc.common
 # Author: Johannes Pohl <johannes.pohl@badaix.de>
 # Co-Author: Szabolcs Hubai <szab.hu@gmail.com>
 # Suggested-by: Tianling Shen <cnsztl@immortalwrt.org>
 START=90
 
 USE_PROCD=1
 
-NAME=snapclient
+NAME=snapserver
 PROG=/usr/bin/$NAME
 PID_FILE=/var/run/$NAME.pid
 
 start_service()
 {
 	config_load "$NAME"
 
 	local autostart
@@ -19,16 +19,16 @@
 	[ "$autostart" -ne "1" ] && logger -t $NAME "Not starting due to disabled autostart" && return 1
 
 	local log_sink opts
 	config_get log_sink "config" log_sink system
 	config_get opts "config" opts ""
 
 	procd_open_instance
 	procd_set_param command $PROG
-	procd_append_param command --logsink $log_sink
+	procd_append_param command --logging.sink $log_sink
 	procd_append_param command $opts
 	procd_set_param pidfile $PID_FILE
 	procd_set_param respawn # use the defaults for respawing crashed process
 	procd_set_param stderr 1
 	procd_set_param stdout 1
 	procd_close_instance
 }
```

Will be there any value in uniforming them? I mean something `*.init.in` and `*.conf.in` files and some transformation magic.
Any thoughts, pointers?

1715173329 commented:
> Will be there any value in uniforming them?

Use single config and init for both services? Based on the config you posted it sounds good to me.

They can be put together simply.

for config:
```
config snapcast 'client'
 	option autostart '0'
 	option log_sink 'system'
 	option opts ''

config snapcast 'server'
 	option autostart '0'
 	option log_sink 'system'
 	option opts ''
```

and init
```
...
if [ "$client_autostart" -eq "1" ]; then
	...
	procd_open_instance client
	...
 	procd_close_instance
fi
...

if [ "$server_autostart" -eq "1" ]; then
	...
	procd_open_instance server
	...
 	procd_close_instance
fi
...
```

xabolcs commented:
> Use single config and init for both services? Based on the config you posted it sounds good to me.

I'm not sure about single config and init files for both services:
* how to name it? :sweat_smile: 
* it's not really common to install both package on the same device: the most common use case is to run `snapserver` on machine1 and run `snapclient` on machine2,3,4,...,n

I thought about single file in the scm (like a template), but generate different files in the binary packages.

1715173329 commented:
> * how to name it?

`snapcast` maybe?

> * it's not really common to install both package on the same device: the most common use case is to run `snapserver` on machine1 and run `snapclient` on machine2,3,4,...,n

you can create a seperated package e.g. `snapcast-config` and let server and client depends on it.

xabolcs commented:
> you can create a seperated package e.g. `snapcast-config` and let server and client depends on it.

Sorry, my imagination level is very low. What should I put in that config package? :sweat_smile: 

Another way: imbuing some transformation magic (`sed`) into the `install` sections is acceptable?

1715173329 commented:
> Sorry, my imagination level is very low. What should I put in that config package? 😅

Hmmm... Maybe there's some misunderstanding... You can take [kcptun](https://github.com/openwrt/packages/blob/master/net/kcptun/Makefile) as an example.

> Another way: imbuing some transformation magic (`sed`) into the `install` sections is acceptable?

I'm fine with it.

xabolcs commented:
Another noob question: how to overcome unrelated build errors? :roll_eyes: 

```bash
touch /openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/gpio-button-hotplug/.configured_68b329da9893e34099c7d8ad5cb9c940
make[2]: *** No rule to make target '/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/linux-6.6.87/.config', needed by '/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/gpio-button-hotplug/.built'.  Stop.
make[2]: Leaving directory '/openwrt/package/kernel/gpio-button-hotplug'
time: package/kernel/gpio-button-hotplug/compile#0.06#0.07#0.13
    ERROR: package/kernel/gpio-button-hotplug failed to build.
make[1]: *** [package/Makefile:189: package/kernel/gpio-button-hotplug/compile] Error 1
make[1]: Leaving directory '/openwrt'
make: *** [/openwrt/include/toplevel.mk:233: package/feeds/local_openwrt_packages/snapcast/compile] Error 2
```

xabolcs commented:
> Hmmm... Maybe there's some misunderstanding... You can take [kcptun](https://github.com/openwrt/packages/blob/master/net/kcptun/Makefile) as an example.

Thanks, it is very helpful!
Going to try that first and as a fallback I would go for `sed`.

1715173329 commented:
> Another noob question: how to overcome unrelated build errors? 🙄
> 
> ```shell
> touch /openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/gpio-button-hotplug/.configured_68b329da9893e34099c7d8ad5cb9c940
> make[2]: *** No rule to make target '/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/linux-6.6.87/.config', needed by '/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/gpio-button-hotplug/.built'.  Stop.
> make[2]: Leaving directory '/openwrt/package/kernel/gpio-button-hotplug'
> time: package/kernel/gpio-button-hotplug/compile#0.06#0.07#0.13
>     ERROR: package/kernel/gpio-button-hotplug failed to build.
> make[1]: *** [package/Makefile:189: package/kernel/gpio-button-hotplug/compile] Error 1
> make[1]: Leaving directory '/openwrt'
> make: *** [/openwrt/include/toplevel.mk:233: package/feeds/local_openwrt_packages/snapcast/compile] Error 2
> ```

`make target/linux/compile -j$(nproc)` if you're building from source

xabolcs commented:
Ahh, that fast upgrading path for Boost (#26382) kills Snapcast again (badaix/snapcast#1367). :confused:

xabolcs commented:
> **xabolcs** force-pushed the `23924-snapcast-add-snapserver-and-snapclient-packages` branch from 489a08eb27cccda412d773cd691f5392c89ac7c5 to 7a6ceead2e87bcb6ea20a8c5d95afbbbbe80cb86

* review suggestions implemented
* compile tested
* not tested runtime yet: the new config parsing system is not ready yet, I think
* had to include the "revert-boost-to-1.87.0" commit :shrug:
  * I'll drop it when the fix is done in upstream
  * this PR will be in draft until then :shrug:

xabolcs commented:
@davidandreoletti , @hunterzero99

First round of addressing review comments and testing on **ipq40xx/generic** is done! :heavy_check_mark: 
Feel free to test / use the current state of this PR!

```bash
root@GL-B1300-62E ~ # logread | grep snap
Sat May  3 19:29:00 2025 user.notice snapclient: Not starting due to disabled autostart
Sat May  3 19:29:00 2025 user.notice snapserver: Not starting due to disabled autostart
root@GL-B1300-62E ~ # vi /etc/config/snapserver 
root@GL-B1300-62E ~ # vi /etc/config/snapclient 
root@GL-B1300-62E ~ # service restart snapserver
Service "restart" not found:
root@GL-B1300-62E ~ # service snapserver restart
root@GL-B1300-62E ~ # service snapclient restart
root@GL-B1300-62E ~ # logread | grep snap
Sat May  3 19:29:00 2025 user.notice snapclient: Not starting due to disabled autostart
Sat May  3 19:29:00 2025 user.notice snapserver: Not starting due to disabled autostart
Sat May  3 19:31:16 2025 user.info snapserver[2735]: Version 0.28.0
Sat May  3 19:31:16 2025 user.info snapserver[2735]: Adding source: pipe:///tmp/snapfifo?name=default
Sat May  3 19:31:16 2025 user.notice snapserver[2735]: Settings file: "//.config/snapserver/server.json"
Sat May  3 19:31:16 2025 user.info snapserver[2735]: Using HTTP host name: GL-B1300-62E
Sat May  3 19:31:16 2025 user.info snapserver[2735]: PcmStream: default, sampleFormat: 48000:16:2
Sat May  3 19:31:16 2025 user.info snapserver[2735]: PipeStream mode: create
Sat May  3 19:31:16 2025 user.info snapserver[2735]: Stream: {"fragment":"","host":"","path":"/tmp/snapfifo","query":{"chunk_ms":"20","codec":"flac","name":"default","sampleformat":"48000:16:2"},"raw":"pipe:////tmp/snapfifo?chunk_ms=20&codec=flac&name=default&sampleformat=48000:16:2","scheme":"pipe"}
Sat May  3 19:31:16 2025 user.info snapserver[2735]: Init - compression level: 2
Sat May  3 19:31:16 2025 user.info snapserver[2735]: Creating TCP acceptor for address: 0.0.0.0, port: 1705
Sat May  3 19:31:16 2025 user.info snapserver[2735]: Creating HTTP acceptor for address: 0.0.0.0, port: 1780
Sat May  3 19:31:16 2025 user.info snapserver[2735]: Creating stream acceptor for address: 0.0.0.0, port: 1704
Sat May  3 19:31:16 2025 user.info snapserver[2735]: Number of threads: 4, hw threads: 4
Sat May  3 19:31:16 2025 user.err snapserver[2735]: Error reading message: End of file, length: 0, ec: asio.misc:2
Sat May  3 19:31:18 2025 user.err snapclient[2801]: PCM device "default" not found
Sat May  3 19:31:18 2025 user.info snapclient[2801]: Version 0.28.0
Sat May  3 19:31:18 2025 user.info snapclient[2801]: (Browser) CACHE_EXHAUSTED
Sat May  3 19:31:20 2025 user.info snapclient[2801]: (Browser) CACHE_EXHAUSTED
Sat May  3 19:31:21 2025 user.info snapclient[2801]: (Browser) CACHE_EXHAUSTED
Sat May  3 19:31:22 2025 user.info snapclient[2801]: (Browser) CACHE_EXHAUSTED
root@GL-B1300-62E ~ # ubus call system board
{
	"kernel": "6.6.87",
	"hostname": "GL-B1300-62E",
	"system": "ARMv7 Processor rev 5 (v7l)",
	"model": "GL.iNet GL-B1300",
	"board_name": "glinet,gl-b1300",
	"rootfs_type": "squashfs",
	"release": {
		"distribution": "OpenWrt",
		"version": "SNAPSHOT",
		"firmware_url": "https://downloads.openwrt.org/",
		"revision": "r29441-a0f45a4f3f",
		"target": "ipq40xx/generic",
		"description": "OpenWrt SNAPSHOT r29441-a0f45a4f3f",
		"builddate": "1746218813"
	}
}
```

@davidandreoletti : this PR will be in draft until the Boost v1.88.0 fix hits upstream. Sorry for that.

davidandreoletti commented:
> @davidandreoletti , @hunterzero99
> 
> 
> 
> First round of addressing review comments and testing on **ipq40xx/generic** is done! :heavy_check_mark: 
> 
> Feel free to test / use the current state of this PR!
> 
> 
> 
> ```bash
> 
> root@GL-B1300-62E ~ # logread | grep snap
> 
> Sat May  3 19:29:00 2025 user.notice snapclient: Not starting due to disabled autostart
> 
> Sat May  3 19:29:00 2025 user.notice snapserver: Not starting due to disabled autostart
> 
> root@GL-B1300-62E ~ # vi /etc/config/snapserver 
> 
> root@GL-B1300-62E ~ # vi /etc/config/snapclient 
> 
> root@GL-B1300-62E ~ # service restart snapserver
> 
> Service "restart" not found:
> 
> root@GL-B1300-62E ~ # service snapserver restart
> 
> root@GL-B1300-62E ~ # service snapclient restart
> 
> root@GL-B1300-62E ~ # logread | grep snap
> 
> Sat May  3 19:29:00 2025 user.notice snapclient: Not starting due to disabled autostart
> 
> Sat May  3 19:29:00 2025 user.notice snapserver: Not starting due to disabled autostart
> 
> Sat May  3 19:31:16 2025 user.info snapserver[2735]: Version 0.28.0
> 
> Sat May  3 19:31:16 2025 user.info snapserver[2735]: Adding source: pipe:///tmp/snapfifo?name=default
> 
> Sat May  3 19:31:16 2025 user.notice snapserver[2735]: Settings file: "//.config/snapserver/server.json"
> 
> Sat May  3 19:31:16 2025 user.info snapserver[2735]: Using HTTP host name: GL-B1300-62E
> 
> Sat May  3 19:31:16 2025 user.info snapserver[2735]: PcmStream: default, sampleFormat: 48000:16:2
> 
> Sat May  3 19:31:16 2025 user.info snapserver[2735]: PipeStream mode: create
> 
> Sat May  3 19:31:16 2025 user.info snapserver[2735]: Stream: {"fragment":"","host":"","path":"/tmp/snapfifo","query":{"chunk_ms":"20","codec":"flac","name":"default","sampleformat":"48000:16:2"},"raw":"pipe:////tmp/snapfifo?chunk_ms=20&codec=flac&name=default&sampleformat=48000:16:2","scheme":"pipe"}
> 
> Sat May  3 19:31:16 2025 user.info snapserver[2735]: Init - compression level: 2
> 
> Sat May  3 19:31:16 2025 user.info snapserver[2735]: Creating TCP acceptor for address: 0.0.0.0, port: 1705
> 
> Sat May  3 19:31:16 2025 user.info snapserver[2735]: Creating HTTP acceptor for address: 0.0.0.0, port: 1780
> 
> Sat May  3 19:31:16 2025 user.info snapserver[2735]: Creating stream acceptor for address: 0.0.0.0, port: 1704
> 
> Sat May  3 19:31:16 2025 user.info snapserver[2735]: Number of threads: 4, hw threads: 4
> 
> Sat May  3 19:31:16 2025 user.err snapserver[2735]: Error reading message: End of file, length: 0, ec: asio.misc:2
> 
> Sat May  3 19:31:18 2025 user.err snapclient[2801]: PCM device "default" not found
> 
> Sat May  3 19:31:18 2025 user.info snapclient[2801]: Version 0.28.0
> 
> Sat May  3 19:31:18 2025 user.info snapclient[2801]: (Browser) CACHE_EXHAUSTED
> 
> Sat May  3 19:31:20 2025 user.info snapclient[2801]: (Browser) CACHE_EXHAUSTED
> 
> Sat May  3 19:31:21 2025 user.info snapclient[2801]: (Browser) CACHE_EXHAUSTED
> 
> Sat May  3 19:31:22 2025 user.info snapclient[2801]: (Browser) CACHE_EXHAUSTED
> 
> root@GL-B1300-62E ~ # ubus call system board
> 
> {
> 
> 	"kernel": "6.6.87",
> 
> 	"hostname": "GL-B1300-62E",
> 
> 	"system": "ARMv7 Processor rev 5 (v7l)",
> 
> 	"model": "GL.iNet GL-B1300",
> 
> 	"board_name": "glinet,gl-b1300",
> 
> 	"rootfs_type": "squashfs",
> 
> 	"release": {
> 
> 		"distribution": "OpenWrt",
> 
> 		"version": "SNAPSHOT",
> 
> 		"firmware_url": "https://downloads.openwrt.org/",
> 
> 		"revision": "r29441-a0f45a4f3f",
> 
> 		"target": "ipq40xx/generic",
> 
> 		"description": "OpenWrt SNAPSHOT r29441-a0f45a4f3f",
> 
> 		"builddate": "1746218813"
> 
> 	}
> 
> }
> 
> ```
> 
> 
> 
> @davidandreoletti : this PR will be in draft until the Boost v1.88.0 fix hits upstream. Sorry for that.

Thank you. I will test it within the next 10d.

PS: You never have to say sorry for work you have done ;-)

xabolcs commented:
> this PR will be in draft until the Boost v1.88.0 fix hits upstream. Sorry for that.

Some progress was made upstream with Boost v1.88.0's compatibility, but it is not yet on `develop`:
![image](https://github.com/user-attachments/assets/7ab4f046-bc4e-4f46-883c-82e3d546be78)

[Those commits](https://github.com/badaix/snapcast/compare/boost-1.88) on `badaix:boost-1.88` branch need to be cleaned a bit.

I'll give the backport a try.

xabolcs commented:
> but GH workflow tests aren't happy!

Only one: [Feeds Package Test Build / Test mips_24kc](https://github.com/openwrt/packages/actions/runs/15833277925/job/44630885343?pr=23956#logs)

```
 Testing package libsqlite3-0 in version 3.49.1 from libs/sqlite3
No pre-test.sh script available
(1/1) Installing libsqlite3-0 (3.49.1-r2)
Bus error (core dumped)
Error: Process completed with exit code 138.
```

I'm going to wait for / do a few runtime test and set this PR ready.

davidandreoletti commented:
> > but GH workflow tests aren't happy!
> 
> Only one: [Feeds Package Test Build / Test mips_24kc](https://github.com/openwrt/packages/actions/runs/15833277925/job/44630885343?pr=23956#logs)
> 
> ```
>  Testing package libsqlite3-0 in version 3.49.1 from libs/sqlite3
> No pre-test.sh script available
> (1/1) Installing libsqlite3-0 (3.49.1-r2)
> Bus error (core dumped)
> Error: Process completed with exit code 138.
> ```
> 
> I'm going to wait for / do a few runtime test and set this PR ready.



> > but GH workflow tests aren't happy!
> 
> Only one: [Feeds Package Test Build / Test mips_24kc](https://github.com/openwrt/packages/actions/runs/15833277925/job/44630885343?pr=23956#logs)
> 
> ```
>  Testing package libsqlite3-0 in version 3.49.1 from libs/sqlite3
> No pre-test.sh script available
> (1/1) Installing libsqlite3-0 (3.49.1-r2)
> Bus error (core dumped)
> Error: Process completed with exit code 138.
> ```
> 
> I'm going to wait for / do a few runtime test and set this PR ready.

I can compile snapcast packages fine but not build an image due errors of this kind. Is it the same case for you (if rebuilding an OpenWRT image) ?

```
(301/516) Installing iptables-mod-ipopt (1.8.10-r1)
ERROR: wget: exited with error 3
ERROR: iptables-mod-ipopt-1.8.10-r1: Input/output error
```

xabolcs commented:
> Here we go again!
> Boost v1.89 breaks Snapcast ...

Upstream commit worked. Compile test, runtime test incoming!

xabolcs commented:
@1715173329 may I ask you for a re-review?
* I implemented the `/etc/config` files
* backported the boost compatibility commits (up to boost 1.89.0)
* rebased to recent packages repo
* run tested on `ath79/nand`

hunterzero99 commented:
> [Test run `mips_24kc` failed](https://github.com/openwrt/packages/actions/runs/17189448183/job/48763011756#logs) with seemingly unrelated error, again? 🤔
> 
> ```
> Testing package libsqlite3-0 in version 3.50.4 from libs/sqlite3
> No pre-test.sh script available
> (1/1) Installing libsqlite3-0 (3.50.4-r1)
> Bus error (core dumped)
> Error: Process completed with exit code 138.
> ```

I _think_ the error there is with the build system configuration for mips_24kc, and nothing to do with snapcast itself.  mips_24kc seems to be the only architecture testing libsqlite3-0 (and failing there).   libsqlite doesn't seem to be a dependency of snapcast.  I'm guessing it's only there because of a mistake -- maybe copying the build system config from another package? 

I'm not familiar with git's test build system, if I was I would offer more suggestions for how to configure it.

1715173329 commented on sound/snapcast/patches/001-v0.29.0-fix-compilation-with-boost-v1.87.0.patch:
```diff
@@ -0,0 +1,29 @@
+From 6ada939adf5c4f7e43d0fbe6ea45492b9564cb8d Mon Sep 17 00:00:00 2001
+From: badaix <johannes.pohl@badaix.de>
+Date: Thu, 19 Dec 2024 22:53:49 +0100
+Subject: [PATCH] Fix compilation with boost v1.87.0
```
I found this:

> Package snapserver is missing dependencies for the following libraries:
> libcrypto.so.3
> libssl.so.3

Please add `libopenssl` to `DEPENDS`, it should resolve the issue.

xabolcs replied:
> I found this:
> > ...
> 
> Please add libopenssl to DEPENDS, it should resolve the issue.

Thank you very much!
Adding that OpenSSL support will be a second phase, later, when Snapcast is already added. Because I plan it to include as a `-full` flavour: I don't want to include OpenSSL where is not installed already.

Until that I planned to stick to v0.28.0 (latest release without OpenSSL support) and do minimal backports (like those boost compatibility commits).



> since these patches are already included in newer release, maybe simply bump PKG_VERSION instead?

I don't understand. This Snapcast package isn't included in OpenWrt yet (sure it's buildable, see #23924 ), why should `PKG_VERSION` be increased before landing?
Sure, after landing, those backport commits should bump that `PKG_VERSION`! :ok_hand: 

Or are you suggesting to include a `PKG_VERSION` bump into each backport commit (to show an example for future backports)? Sure, I'll buy that reasoning! :slightly_smiling_face:

xabolcs commented on sound/snapcast/files/snapclient.conf:
```diff
@@ -0,0 +1,4 @@
+config snapclient 'config'
+	option autostart '0'
```
> would be better to use `enabled` instead of `autostart`

I'll think about it! Thanks for the suggestion. Will decide later.

xabolcs commented on sound/snapcast/Makefile:
```diff
@@ -0,0 +1,107 @@
+  $(call Package/snapcast/Default)
+  TITLE+=Snapserver
+  DEPENDS+=+libvorbis
+  HIDDEN:=1
```
> hmmm any reason to hide this package from menuconfig?

I don't know! All these code is legacy to me, and it's my first OpenWrt package ever! :grinning:
Will check how does it look when its not hidden!

1715173329 replied:
sorry I didn't notice that menu. I'm okay with it then.

xabolcs commented:
> **xabolcs** marked this pull request as **draft**

A few things popped up with the latest review, so I have a little homework now, but there is already an `r+` too! :pray: 

> otherwise LGTM

xabolcs commented:
> **xabolcs** marked this pull request as ready for review **now**

Run tested again on **ath79/nand**: GL-AR300M
Changes:
* renamed config option to `autostart` to `enabled`
* adjusted related log message

```bash

BusyBox v1.37.0 (2025-08-24 11:22:09 UTC) built-in shell (ash)

  _______                     ________        __
 |       |.-----.-----.-----.|  |  |  |.----.|  |_
 |   -   ||  _  |  -__|     ||  |  |  ||   _||   _|
 |_______||   __|_____|__|__||________||__|  |____|
          |__| W I R E L E S S   F R E E D O M
 -----------------------------------------------------
 OpenWrt SNAPSHOT, r30846-26c6fb671e
 -----------------------------------------------------
 | Machine: GL.iNet GL-AR300M (NAND)                 |
 | Uptime: 0d, 00:03:14                              |
 | Load: 0.51 0.56 0.25                              |
 | Flash: total: 100.8MB, free: 90.2MB, used: 6%     |
 | Memory: total: 116.3MB, free: 92.9MB, used: 20%   |
 | WAN:                                              |
 | LAN: 192.168.1.1, leases: 0                       |
 | radio0: mode: sta, ssid: OpenWrt, channel: 1, conn: 0 |
 -----------------------------------------------------
=== WARNING! =====================================
There is no root password defined on this device!
Use the "passwd" command to set up a new password
in order to prevent unauthorized SSH logins.
--------------------------------------------------
root@GL-AR300M-F5E-NAND ~ # logread -e snap
Tue Sep  2 12:51:19 2025 user.notice snapserver: Disabled in /etc/config/snapserver
Tue Sep  2 12:51:24 2025 user.info snapclient[1716]: Version 0.28.0
Tue Sep  2 12:51:24 2025 user.info snapclient[1716]: Resolving host IP for: snapcast
Tue Sep  2 12:51:31 2025 user.err snapclient[1716]: Failed to resolve host 'snapcast', error: Host not found (non-authoritative), try again later
Tue Sep  2 12:51:31 2025 user.err snapclient[1716]: Error: Host not found (non-authoritative), try again later
Tue Sep  2 12:51:32 2025 user.info snapclient[1716]: Resolving host IP for: snapcast
Tue Sep  2 12:51:37 2025 user.err snapclient[1716]: Failed to resolve host 'snapcast', error: Host not found (non-authoritative), try again later
Tue Sep  2 12:51:37 2025 user.err snapclient[1716]: Error: Host not found (non-authoritative), try again later
Tue Sep  2 12:51:38 2025 user.info snapclient[1716]: Resolving host IP for: snapcast
Tue Sep  2 12:51:38 2025 user.info snapclient[1716]: Connecting to 192.168.xxx.yyy:1704
Tue Sep  2 12:51:38 2025 user.notice snapclient[1716]: Connected to 192.168.xxx.yyy
Tue Sep  2 12:51:38 2025 user.info snapclient[1716]: My MAC: "e4:95:de:ad:bf:ff", socket: 8
Tue Sep  2 12:51:39 2025 user.info snapclient[1716]: ServerSettings - buffer: 1000, latency: 0, volume: 100, muted: 0
Tue Sep  2 12:51:39 2025 user.info snapclient[1716]: Codec: flac, sampleformat: 44100:16:2
Tue Sep  2 12:51:39 2025 user.info snapclient[1716]: Player name: alsa, device: default, description: Default Audio Device, idx: 1, sharing mode: unspecified, parameters: <none>
Tue Sep  2 12:51:39 2025 user.info snapclient[1716]: Mixer mode: software, parameters: <none>
Tue Sep  2 12:51:39 2025 user.info snapclient[1716]: Sampleformat: 44100:16:2, stream: 44100:16:2
Tue Sep  2 12:51:39 2025 user.info snapclient[1716]: Using default buffer_time: 80 ms, default fragments: 4
Tue Sep  2 12:51:39 2025 user.info snapclient[1716]: Period time too small, changing from 20000 to 21333
Tue Sep  2 12:51:39 2025 user.info snapclient[1716]: PCM name: default, sample rate: 44100 Hz, channels: 2, buffer time: 85328 us, periods: 4, period time: 21333 us, period frames: 940
Tue Sep  2 12:51:40 2025 user.info snapclient[1716]: diff to server [ms]: 1.16e+10
Tue Sep  2 12:52:37 2025 user.info snapclient[1716]: pBuffer->full() && (abs(median_) > 2): 2219
Tue Sep  2 12:52:47 2025 user.info snapclient[1716]: pBuffer->full() && (abs(median_) > 2): 3490
Tue Sep  2 12:52:58 2025 user.info snapclient[1716]: pBuffer->full() && (abs(median_) > 2): 2681
Tue Sep  2 12:53:09 2025 user.info snapclient[1716]: pBuffer->full() && (abs(median_) > 2): 3214
Tue Sep  2 12:53:19 2025 user.info snapclient[1716]: pBuffer->full() && (abs(median_) > 2): 2903
Tue Sep  2 12:53:30 2025 user.info snapclient[1716]: pBuffer->full() && (abs(median_) > 2): 2928
Tue Sep  2 12:53:41 2025 user.info snapclient[1716]: pBuffer->full() && (abs(median_) > 2): 3017
Tue Sep  2 12:53:52 2025 user.info snapclient[1716]: pBuffer->full() && (abs(median_) > 2): 3083
Tue Sep  2 12:54:02 2025 user.info snapclient[1716]: pBuffer->full() && (abs(median_) > 2): 2952
Tue Sep  2 12:54:13 2025 user.info snapclient[1716]: pBuffer->full() && (abs(median_) > 2): 2915
Tue Sep  2 12:54:24 2025 user.info snapclient[1716]: pBuffer->full() && (abs(median_) > 2): 3006
Tue Sep  2 12:54:34 2025 user.info snapclient[1716]: pBuffer->full() && (abs(median_) > 2): 2876
root@GL-AR300M-F5E-NAND ~ # cat /etc/config/snapserver 
config snapserver 'config'
	option enabled '0'
	option log_sink 'system'
	option opts ''
root@GL-AR300M-F5E-NAND ~ # cat /etc/config/snapclient 
config snapclient 'config'
	option enabled '1'
	option log_sink 'system'
	option opts '--host snapcast'
root@GL-AR300M-F5E-NAND ~ # 
```

xabolcs commented on sound/snapcast/Config.in:
```diff
@@ -0,0 +1,15 @@
+		prompt "Snapserver"
+		help
+			Snapcast server
+		default n
```
May I ask for suggestion @neheb , @1715173329?
Will this `default n`  trigger a build for this `snapserver` on buildbots? :thinking: 

My points:
* this is the old [badaix/snapos](https://github.com/badaix/snapos/blob/29a5daebb0513faa825b12474125c0c2dd66c2ee/openwrt/snapcast/Makefile#L74C1-L90C6) way
* building manually it looks good to me to selecting `snapclient`, but leaving `snapserver` out by default
* but I surely want to build both `snapclient` and `snapserver` package by the buildbots!

1715173329 replied:
> Will this `default n` trigger a build for this `snapserver` on buildbots? 🤔

Unfortunately no since it's exported as an option, not a package.

This menu makes things a bit complex tbh.
Maybe you can add `default m if BUILDBOT`

or create a meta package:
```patch
diff --git a/sound/snapcast/Config.in b/sound/snapcast/Config.in
deleted file mode 100644
index 29ad05a0fd8e..000000000000
--- a/sound/snapcast/Config.in
+++ /dev/null
@@ -1,15 +0,0 @@
-menu "Select Snapcast Options"
-	depends on PACKAGE_snapcast
-	comment "Choose server or client"
-	config PACKAGE_snapserver
-		prompt "Snapserver"
-		help
-			Snapcast server
-		default n
-
-	config PACKAGE_snapclient
-		prompt "Snapclient"
-		help
-			Snapcast client
-		default y
-endmenu
diff --git a/sound/snapcast/Makefile b/sound/snapcast/Makefile
index be9be1505f8f..0ba6621e7ba4 100644
--- a/sound/snapcast/Makefile
+++ b/sound/snapcast/Makefile
@@ -35,6 +35,7 @@ endef
 
 define Package/snapcast
   $(call Package/snapcast/Default)
+  MENU:=1
   TITLE+=packages
 endef
 
@@ -48,15 +49,13 @@ endef
 define Package/snapserver
   $(call Package/snapcast/Default)
   TITLE+=Snapserver
-  DEPENDS+=+libvorbis
-  HIDDEN:=1
+  DEPENDS+=snapcast +libvorbis
 endef
 
 define Package/snapclient
   $(call Package/snapcast/Default)
   TITILE+=Snapclient
-  DEPENDS+=+libvorbisidec
-  HIDDEN:=1
+  DEPENDS+=snapcast +libvorbisidec
 endef
 
 define Package/snapserver/description
@@ -69,10 +68,6 @@ define Package/snapclient/description
   Snapcast client
 endef
 
-define Package/snapcast/config
-  source "$(SOURCE)/Config.in"
-endef
-
 define Package/snapserver/conffiles
 /etc/config/snapserver
 /etc/snapserver.conf
@@ -82,6 +77,10 @@ define Package/snapclient/conffiles
 /etc/config/snapclient
 endef
 
+define Package/snapcast/install
+	true
+endef
+
 define Package/snapserver/install
 	$(INSTALL_DIR) $(1)/usr/bin
 	$(INSTALL_BIN) $(PKG_BUILD_DIR)/bin/snapserver $(1)/usr/bin/
```

not grouping them would be the easiest way to go:
```patch
diff --git a/sound/snapcast/Makefile b/sound/snapcast/Makefile
index be9be1505f8f..10d31d56c4ec 100644
--- a/sound/snapcast/Makefile
+++ b/sound/snapcast/Makefile
@@ -30,16 +30,6 @@ define Package/snapcast/Default
 endef
 
 define Package/snapcast/description/Default
-  Synchronous audio player
-endef
-
-define Package/snapcast
-  $(call Package/snapcast/Default)
-  TITLE+=packages
-endef
-
-define Package/snapcast/description
-  $(call Package/snapcast/description/Default)
   Snapcast is a multi-room client-server audio player,
   where all clients are time synchronized with the server
   to play perfectly synced audio
@@ -49,14 +39,12 @@ define Package/snapserver
   $(call Package/snapcast/Default)
   TITLE+=Snapserver
   DEPENDS+=+libvorbis
-  HIDDEN:=1
 endef
 
 define Package/snapclient
   $(call Package/snapcast/Default)
   TITILE+=Snapclient
   DEPENDS+=+libvorbisidec
-  HIDDEN:=1
 endef
 
 define Package/snapserver/description
@@ -69,10 +57,6 @@ define Package/snapclient/description
   Snapcast client
 endef
 
-define Package/snapcast/config
-  source "$(SOURCE)/Config.in"
-endef
-
 define Package/snapserver/conffiles
 /etc/config/snapserver
 /etc/snapserver.conf
@@ -99,4 +83,3 @@ endef
 
 $(eval $(call BuildPackage,snapserver))
 $(eval $(call BuildPackage,snapclient))
-$(eval $(call BuildPackage,snapcast))
```

xabolcs replied:
Thank you very much for all the hints! :pray: 
I'll try them out!

xabolcs commented on sound/snapcast/Makefile:
```diff
@@ -0,0 +1,102 @@
+define Package/snapserver
+  $(call Package/snapcast/Default)
+  TITLE+=Snapserver
+  DEPENDS+=snapcast +libvorbis
```
> in this case you have to add `Package/snapcast/install` ...

Added! Thanks!

davidandreoletti commented:
@xabolcs LGTM.

1715173329 commented on sound/snapcast/Makefile:
```diff
@@ -0,0 +1,106 @@
+
+define Package/snapcast/Default
+  SECTION:=sound
+  CATEGORY:=Sound
```
```suggestion
  CATEGORY:=Sound
  SUBMENU:=Snapcast
```

This might be the best choice for "grouping" 2 packages, without creating another dummy package.

What do you think?

xabolcs replied:
> What do you think?


This would be my first package in `openwrt/packages` so my opinion is almost to none.
I'll just follow what you (and the other committers) suggests. :sweat_smile: 

As I can see this dummy `snapcast` package definition in `Makefile` it looks to me as a shortcut or a template to avoid repeating things for `snapclient` and `snapserver`.
I think this because of the `$(call Package/snapcast/*) calls.

I see something like that in `mpd`'s configuration: `Package/mpd/Default` contains all the reusable thing and there is to (variant) package of `mpd`.
And if I do `scripts/feeds install mpd` then all the dependencies and all `mpd` variant got installed into the build system.

It's OK to me to forget the grouping, but in my next PR (after I learn packaging) I plan to upgrade to latest Snapcast and plan to deal with OpenSSL support: I'd like to introduce `full` (with OpenSSL) and `mini` (without OpenSSL) (and of course some `snap{cast,server,client}-avahi-service` to deal with the Avahi dependency) variants which would be at least 2x2 = 4 different packages.
Would be fine to list all those `snap*` packages in the top level of "Sound" category?

As I wrote already, I am a package maintainer newbie, so I prefer what you suggests. :slightly_smiling_face:

xabolcs replied:
> _provides_ the `mpd` package.

Ahh ohh ... so I won't have such a shortcut for Snapcast?
I mean: when I do `scripts/feeds install snapcast` I'd like to have all `snap*` installed in the build system. :thinking: 

I don't wanna do `scripts/feeds install snapserver` and `scripts/feeds install snapclient` (and `scripts/feeds install snapseserver-avahi-service`, once I got there) :confused:

xabolcs replied:
Thank you very much! Looks like I really have to learn a lot!

xabolcs replied:
> > Would be fine to list all those `snap*` packages in the top level of "Sound" category?
> 
> Yes for sure ;)

Went that way, dropped (sub) menu.
Thanks for the suggestion.

xabolcs commented:
PR is ready again!
* dropped the grouping
* extended description

<img width="1086" height="433" alt="Screenshot from 2025-09-07 23-13-25" src="https://github.com/user-attachments/assets/5a7e4364-456a-4bf9-88c7-6c6d842fd344" />
<img width="1085" height="434" alt="Screenshot from 2025-09-07 23-13-55" src="https://github.com/user-attachments/assets/a714f8f9-5d08-4980-a528-d843dd36ba65" />

GeorgeSapkin commented:
@xabolcs I meant a version check or an equivalent.

xabolcs commented:
There is a `--version` for `snapclient`, too!

```bash
root@HT-TM02-43C ~ # snapclient --version
snapclient v0.28.0
Copyright (C) 2014-2022 BadAix (snapcast@badaix.de).
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by Johannes M. Pohl and contributors <https://github.com/badaix/snapcast/graphs/contributors>.

````

But I still don't know where to start. Do you have some pointers where to start looking?
Did I asked relevant questions in my previous comment, at least? Did I understand your suggestion correctly?

GeorgeSapkin commented:
If there are just two binaries and their names match those of their respective packages, and they both support a `--version` argument, it should be enough to add `test.sh` in the package root directory:

```bash
#!/bin/sh

"$1" --version | grep -F "$PKG_VERSION"
```

The script is used by CI in the [Test via Docker container ](https://github.com/openwrt/packages/actions/runs/17534179291/job/49795054175?pr=23956#step:18:1835)step.

xabolcs commented on sound/snapcast/Makefile:
```diff
@@ -0,0 +1,100 @@
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=snapcast
+PKG_VERSION:=0.28.0
```
Hi @GeorgeSapkin !

> The latest release is [0.32.3](https://github.com/badaix/snapcast/releases/tag/v0.32.3). Why not use that one?

It started with 0.29.0 where OpenSSL support introduced and [I got problems](https://github.com/openwrt/packages/pull/23956#issuecomment-2591962182) with it (back in January 2025):

> Well in the meantime Snapcast 0.29.0 came out which has HTTPS support, but I'm so noob I don't know how to add those dependencies! 🙈
> I mean:
> ```bash
> Package snapserver is missing dependencies for the following libraries:
> libcrypto.so.3
> libssl.so.3
> ```
> Sure, I could stick with 0.28.0. 🤷

At that point of time I decided to stick with 0.28.0 and deal with this OpenSSL support later with a followup PR where I introduce different flavours: `snap*-mini` and `snap*-full`.
And I also [want to](https://github.com/openwrt/packages/pull/23956#issuecomment-2661456172) make the Avahi dependency optional: badaix/snapos#4

And then, last week @1715173329 [suggested](https://github.com/openwrt/packages/pull/23956#discussion_r2311924959) how to solve the problem
> I found this:
>
> ```bash
> Package snapserver is missing dependencies for the following libraries:
> libcrypto.so.3
> libssl.so.3
> ```
> Please add libopenssl to DEPENDS, it should resolve the issue.

And now we are here.

In summary:
* I'd like to include `snapcast` before OpenWrt 25.xx branches off (and hopefully backport to 24.10, @hunterzero99 had a [success build for once](https://github.com/openwrt/packages/pull/23956#issuecomment-2711323876))
* with that in mind I choose to stick to 0.28.0 and backport compatibility fixes
  * because I don't know how to be a maintainer in `openwrt/packages` and this would be my first package
* and after I learn how to deal with those things above, I update to latest tagged Snapcast or even to specific `develop` commit ([see talk](https://github.com/openwrt/packages/pull/23956#pullrequestreview-2607569304) with @neheb)

If you have time and energy to coordinate me with these problems (disabling the OpenSSL support is on upstream `develop`, see badaix/snapcast#1427) I happily go this way! :ok_hand: :pray:

GeorgeSapkin replied:
Thanks for the summary. I guess if everybody else is OK with this, then I'm OK.

xabolcs commented:
> Please do not add `Suggested-by` tag to real files, leaving them in the commit message is good enough.

Done!

1715173329 commented:
Merged, thank you!

xabolcs commented:
Sorry for creating faillog for `snapcast` since the beginning!
I started to fix that PulseAudio dependency.
I'm going to cut that (and while at it, the Avahi) dependency out.

I'm very sorry!
